code/eyetracking/eyetrackingrun.py

```diff
@@ -0,0 +1,961 @@
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import re
+import os
+import json
+from pathlib import Path
+import string
+from collections import defaultdict
+from typing import Optional, List, Tuple, Union
+
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from pyedfread import edf, edfread
+
+
+# EyeLink calibration coordinates from https://www.sr-research.com/calibration-coordinate-calculator/
+EYELINK_CALIBRATION_COORDINATES = [
+    (400, 300), (400, 51), (400, 549), (48, 300), (752, 300), (48, 51), (752, 51), (48, 549), (752, 549), (224, 176), (576, 176), (224, 424), (576, 424)
+]
+
+EYE_CODE_MAP = defaultdict(
+    lambda: "unknown", {"R": "right", "L": "left", "RL": "both"}
+)
+RIGHT_EYE_COLUMNS = {
+    "time": "eye_timestamp",
+    "gx_right": "eye1_x_coordinate",
+    "gy_right": "eye1_y_coordinate",
+    "pa_right": "eye1_pupil_size",
+    "px_right": "eye1_pupil_x_coordinate",
+    "py_right": "eye1_pupil_y_coordinate",
+    "hx_right": "eye1_head_x_coordinate",
+    "hy_right": "eye1_head_y_coordinate",
+    "rx": "screen_pixel_per_degree_x",
+    "ry": "screen_pixel_per_degree_y",
+    "gxvel_right": "eye1_x_velocity",
+    "gyvel_right": "eye1_x_velocity",
+    "hxvel_right": "eye1_head_x_velocity",
+    "hyvel_right": "eye1_head_y_velocity",
+    "rxvel_right": "eye1_raw_x_velocity",
+    "ryvel_right": "eye1_raw_y_velocity",
+    "fgxvel": "eye1_fast_gaze_x_velocity",
+    "fgyvel": "eye1_fast_gaze_y_velocity",
+    "fhxyvel": "eye1_fast_head_x_velocity",
+    "fhyvel": "eye1_fast_head_y_velocity",
+    "frxyvel": "eye1_fast_raw_x_velocity",
+    "fryvel": "eye1_fast_raw_y_velocity",
+}
+
+
+class EyeTrackingRun:
+    """
+    Class representing an instance of eye tracking data.
+
+    Parameters
+    ----------
+    session : int
+        Session ID.
+    task_name : str
+        Task name.
+    participant : int
+        Participant ID.
+    samples : pd.DataFrame
+        DataFrame containing eye tracking samples.
+    events : pd.DataFrame
+        DataFrame containing eye tracking events.
+    messages : pd.DataFrame
+        DataFrame containing eye tracking messages.
+    message_first_trigger : str
+        Message that signals the first trigger of the run.
+    screen_resolution : Tuple[int, int]
+        Screen resolution as a tuple of two integers.
+
+    pe : str, optional
+        Phase encoding direction (default is an empty string).
+
+    Examples
+    --------
+    This class can be instantiated as follows:
+
+    ```python
+    et_run = EyeTrackingRun(
+        session=1,
+        task_name="rest",
+        participant=001,
+        samples=samples_df,
+        events=events_df,
+        messages=messages_df,
+        message_first_trigger="start",
+        screen_resolution=(800, 600),
+        messages_start_fixation="fixation_start",
+        messages_stop_fixation="fixation_stop",
+        pe="",
+    )
+    ```
```

Comment:
Better as a doctest:

```suggestion
    >>> et_run = EyeTrackingRun(
    ...     session=1,
    ...     task_name="rest",
    ...     participant=001,
    ...     samples=samples_df,
    ...     events=events_df,
    ...     messages=messages_df,
    ...     message_first_trigger="start",
    ...     screen_resolution=(800, 600),
    ...     messages_start_fixation="fixation_start",
    ...     messages_stop_fixation="fixation_stop",
    ...     pe="",
    ... )
```